### ISSUE TITLE
"tried to advance finished generator"

### ISSUE THREAD

**Issue by JackCA** [machty/ember-concurrency#138]
I'm currently seeing the ["tried to advance finished generator"](https://github.com/machty/ember-concurrency/blob/aae2822ee746600d553ada543c57e47890ddb4aa/addon/-task-instance.js#L412) exception being thrown in my CircleCI tests but not locally.

It's difficult for me debug due to it being remote and I can provide more debugging information if needed, but generally, how would I be hitting this as a user?

**Comment by JackCA** [comment]
... and I've figured out the issue:

I had attempted to go against the [`events` docs](http://ember-concurrency.com/#/docs/events) and tried to avoid the while loop and use recursion:

```javascript
foo: task(function * () {
  yield waitForEvent(this.$(), 'click')
  this.get('foo').perform()
})
```

this is what was triggering the problem -- switching to `while (true)` worked.

I'm not sure if this is intended behavior or something that could/should be added to the docs.

**Comment by machty** [comment]
More people are seeing this, gonna reopen.

**Comment by machty** [comment]
@JackCA were you seeing this error on a `.restartable()` task? I think that's the only case where the recursion will bite: because new EC scheduler tries to be as synchronous as possible, when you try and recursively perform a restartable task, it'll synchronously try to cancel the task, which isn't possible if the `.perform()` is _within the call stack_ of the task. I basically didn't have a good idea as to how to solve this problem so I just made it an error for the current version, but we probably need to find a better solution.

**Comment by JackCA** [comment]
@machty I believe it was (which means my example above was inaccurate)

 I'm not sure if allowing recursion like that is important (create a new stack?), but it's possible that the thrown message could be improved to communicate to users why this might happen and how to solve. Are there other ways this exception could be hit?

cc @tzellman

**Comment by tzellman** [comment]
Cool. #148 is essentially a duplicate of this issue. Feel free to close that one as such if you like @machty.

**Comment by courajs** [comment]
Yeah, this happens when a task cancels itself synchronously:
```js
foo: task(function * () {
  this.get('foo').cancelAll();
})
```

**Comment by mmun** [comment]
@courajs That's what I experienced as well.

**Comment by machty** [comment]
should be fixed by 8aec0b3
